File: 059.py
#   059
#   Spiral Matrix II
#   2015-12-16
#####################################################
class Solution(object):
    def generateMatrix(self, n):
        """
        :type n: int
        :rtype: List[List[int]]
        """
        m = [[0]*n for i in range(n)]
        self.fill(m, 0, n-1, 0)
        return m

    def fill(self, m, p, q, count):
        if p > q:
            return
        elif p == q:
            count+=1
            m[p][p] = count
        for i in range(p, q):
            count += 1
            m[p][i] = count
        for i in range(p, q):
            count += 1
            m[i][q] = count
        for i in range(q, p, -1):
            count += 1
            m[q][i] = count
        for i in range(q, p, -1):
            count += 1
            m[i][p] = count
        self.fill(m, p+1, q-1, count)

#####################################################
sol = Solution()

n = 3
m = sol.generateMatrix(n)
for v in m:
    print v

File: leetcode/leetcode/059.cpp
//
//  059.cpp
//  leetcode
//
//  Created by Hang Zhang on 12/15/15.
//  Copyright (c) 2015 Hilings Studio. All rights reserved.
//

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        vector<vector<int>> m;
        for (int i = 0; i < n; i++) {
            m.push_back(vector<int> (n, 0));
        }
        fill(m, 0, n-1, 0);
        return m;
    }
    void fill(vector<vector<int>> &m, int p, int q, int count) {
        if (p == q) {
            m[p][p] = ++count;
            return;
        } else if (p > q) {
            return;
        }

        for (int i = p; i < q; i++) {
            m[p][i] = ++count;
        }
        for (int i = p; i < q; i++) {
            m[i][q] = ++count;
        }
        for (int i = q; i > p; i--) {
            m[q][i] = ++count;
        }
        for (int i = q; i > p; i--) {
            m[i][p] = ++count;
        }
        fill(m, p+1, q-1, count);
    }
};

int main(int arg, char *argv[]) {
    // insert code here...
    cout << "LeetCode 059 Spiral Matrix II, C++ ...\n\n";

    int n = 3;
    Solution sol;
    vector<vector<int>> m = sol.generateMatrix(n);
    for (int i = 0; i < m.size(); i++) {
        for (int j = 0; j < m[i].size(); j++) {
            cout << m[i][j] << ' ';
        }
        cout << '\n';
    }

    return 0;
}